### ISSUE TITLE
Async bootstrap dropping messages

### ISSUE THREAD

**Issue by DPGrev** [zendesk/maxwell#1418]
We have a few large tables which we bootstrapped. In the case of the async bootstrap, we lost all messages for that table while the table was bootstrapping. 

We saw the following error in the logs after the bootstrapping was done:

```
20:35:19,566 INFO  SynchronousBootstrapper - bootstrapping started for <<TABLE NAME>>
21:37:51,298 INFO  SynchronousBootstrapper - bootstrapping ended for #2958 <<TABLE NAME>>

com.mysql.cj.jdbc.exceptions.CommunicationsException: Communications link failure

The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server.
	at com.mysql.cj.jdbc.exceptions.SQLError.createCommunicationsException(SQLError.java:174)
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:64)
	at com.mysql.cj.jdbc.ConnectionImpl.createNewIO(ConnectionImpl.java:827)
	at com.mysql.cj.jdbc.ConnectionImpl.<init>(ConnectionImpl.java:447)
	at com.mysql.cj.jdbc.ConnectionImpl.getInstance(ConnectionImpl.java:237)
	at com.mysql.cj.jdbc.NonRegisteringDriver.connect(NonRegisteringDriver.java:199)
	at java.sql.DriverManager.getConnection(DriverManager.java:664)
	at java.sql.DriverManager.getConnection(DriverManager.java:247)
	at snaq.db.ConnectionPool.create(ConnectionPool.java:256)
	at snaq.db.ConnectionPool.create(ConnectionPool.java:71)
	at snaq.util.ObjectPool.checkOut(ObjectPool.java:377)
	at snaq.db.ConnectionPool.getConnection(ConnectionPool.java:385)
	at com.zendesk.maxwell.bootstrap.BootstrapController.getIncompleteTasks(BootstrapController.java:88)
	at com.zendesk.maxwell.bootstrap.BootstrapController.work(BootstrapController.java:57)
	at com.zendesk.maxwell.util.RunLoopProcess.runLoop(RunLoopProcess.java:34)
	at com.zendesk.maxwell.MaxwellContext.lambda$getBootstrapController$0(MaxwellContext.java:395)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.mysql.cj.exceptions.CJCommunicationsException: Communications link failure
```

Maxwell continues running without actually sending the events that it stored in the queue whilst bootstrapping. Resulting in a loss of data.

I think this system should be more robust and possibly, maybe we can have the connection pool retry getting a connection before actually throwing an error?

**Comment by osheroff** [comment]
yeah some retry logic in there is a good idea.  At a certain point we'd really have to crash though.  But I'll put in some retry logic for a communications exception.  

just thinking as I write; is there a case where a communications exception means we *want* to crash?  Like if we're nominating a new master database, would we want to crash out and restart?  will have to think about corner cases...

**Comment by DPGrev** [comment]
I think we would want to crash after a certain amount of retries, but this should be configurable. If the user would want to opt for crashing directly and maybe changing master then retries should be set to 0. 

Also, I think the underlying connection pool package is not maintained anymore. Maybe a connection pool that has retries build in could be used? So that part is maintained outside of this repo.

**Comment by osheroff** [comment]
fixed this crash in 1.25.0
